bemani/backend/iidx/cannonballers.py

```diff
@@ -1,15 +1,2018 @@
+            detail.set_attribute('hair', str(qpro.get_int('hair')))
+            detail.set_attribute('face', str(qpro.get_int('face')))
+            detail.set_attribute('body', str(qpro.get_int('body')))
+            detail.set_attribute('hand', str(qpro.get_int('hand')))
+
+        return root
+
+    def handle_IIDX25ranking_entry_request(self, request: Node) -> Node:
+        extid = int(request.attribute('iidxid'))
+        courseid = int(request.attribute('coid'))
+        chart = int(request.attribute('clid'))
+        course_type = int(request.attribute('regist_type'))
+        clear_status = self.game_to_db_status(int(request.attribute('clr')))
+        pgreats = int(request.attribute('pgnum'))
+        greats = int(request.attribute('gnum'))
+
+        if course_type == 0:
+            index = self.COURSE_TYPE_INTERNET_RANKING
+        elif course_type == 1:
+            index = self.COURSE_TYPE_SECRET
+        else:
+            raise Exception('Unknown registration type for course entry!')
+
+        userid = self.data.remote.user.from_extid(self.game, self.version, extid)
+        if userid is not None:
+            # Update achievement to track course statistics
+            self.update_course(
+                userid,
+                index,
+                courseid,
+                chart,
+                clear_status,
+                pgreats,
+                greats,
+            )
+
+        # We should return the user's position, but its not displayed anywhere
+        # so fuck it.
+        root = Node.void('IIDX25ranking')
+        root.set_attribute('anum', '1')
+        root.set_attribute('jun', '1')
+        return root
+
+    def handle_IIDX25ranking_classicentry_request(self, request: Node) -> Node:
+        extid = int(request.attribute('iidx_id'))
+        courseid = int(request.attribute('course_id'))
+        coursestyle = int(request.attribute('play_style'))
+        clear_status = self.game_to_db_status(int(request.attribute('clear_flg')))
+        pgreats = int(request.attribute('pgnum'))
+        greats = int(request.attribute('gnum'))
+
+        userid = self.data.remote.user.from_extid(self.game, self.version, extid)
+        if userid is not None:
+            # Update achievement to track course statistics
+            self.update_course(
+                userid,
+                self.COURSE_TYPE_CLASSIC,
+                courseid,
+                coursestyle,
+                clear_status,
+                pgreats,
+                greats,
+            )
+
+        return Node.void('IIDX25ranking')
+
+    def handle_IIDX25music_crate_request(self, request: Node) -> Node:
+        root = Node.void('IIDX25music')
+        attempts = self.get_clear_rates()
+
+        all_songs = list(set([song.id for song in self.data.local.music.get_all_songs(self.game, self.music_version)]))
+        for song in all_songs:
+            clears = []
+            fcs = []
+
+            for chart in [0, 1, 2, 3, 4, 5]:
```

Comment:
I think the mistake was that IIDX originally did not have any indirection in charts (jubeat might not either). We will need to introduce that in the existing IIDX versions to be correct.